File: src/1028.从先序遍历还原二叉树.py
#
# @lc app=leetcode.cn id=1028 lang=python3
#
# [1028] 从先序遍历还原二叉树
#
import unittest
from utils import TreeNode, Codec
# @lc code=start
from typing import Generator, Tuple


class Solution:
    def parse_string(self, s: str) -> Generator[Tuple[int, int], None, None]:
        n = len(s)
        i = 0
        #
        while i < n:
            # find depth
            depth = 0
            while i < n and s[i] == '-':
                depth += 1
                i += 1

            # find number
            number = 0
            while i < n and s[i].isdigit():
                number = number * 10 + int(s[i])
                i += 1
            yield (depth, number)

    def recoverFromPreorder(self, S: str) -> TreeNode:
        nodes = self.parse_string(S)
        stack = []

        super_root = TreeNode(-1)
        stack.append((-1, super_root))
        for (depth, val) in nodes:
            # pop until depth == stack[-1].depth + 1
            while depth <= stack[-1][0]:
                stack.pop()
            # insert to left or right
            node = TreeNode(val)
            last_node = stack[-1][1]
            if last_node.left is None:
                last_node.left = node
            else:
                last_node.right = node
            # push to stack
            stack.append((depth, node))
        return super_root.left

# @lc code=end


class Test(unittest.TestCase):
    def test(self):
        def f(s, ans):
            t1 = Solution().recoverFromPreorder(s)
            t2 = Codec().deserialize(ans)
            self.assertEqual(t1, t2)

        f(
            "1-2--3--4-5--6--7",
            '[1,2,5,3,4,6,7]'
        )
        f(
            "1-2--3---4-5--6---7",
            '[1,2,5,3,null,6,null,4,null,7]'
        )
        f(
            "1-401--349---90--88",
            '[1,401,null,349,88,90]'
        )


if __name__ == "__main__":
    unittest.main()

File: src/1028.从先序遍历还原二叉树.rs
/*
 * @lc app=leetcode.cn id=1028 lang=rust
 *
 * [1028] 从先序遍历还原二叉树
 */
mod utils;
use utils::TreeNode;

struct Solution;

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

struct Parse {
    chars: Vec<u8>,
    i: usize,
}

impl Parse {
    pub fn from(s: String) -> Self {
        Self {
            chars: s.into_bytes(),
            i: 0,
        }
    }
}

impl std::iter::Iterator for Parse {
    type Item = (i32, i32);
    fn next(&mut self) -> Option<Self::Item> {
        if self.i == self.chars.len() {
            return None;
        }
        // find dashes
        let mut depth = 0;
        while self.i < self.chars.len() && self.chars[self.i] == ('-' as u8) {
            depth += 1;
            self.i += 1;
        }
        let mut number = 0;
        while self.i < self.chars.len() && self.chars[self.i].is_ascii_digit() {
            number = number * 10 + (self.chars[self.i] - '0' as u8) as i32;
            self.i += 1;
        }

        Some((depth, number))
    }
}

impl Solution {
    pub fn recover_from_preorder(s: String) -> Option<Rc<RefCell<TreeNode>>> {
        let super_root = Rc::new(RefCell::new(TreeNode::new(0)));
        let mut stack = vec![(-1, super_root.clone())];
        //
        for (depth, val) in Parse::from(s) {
            while depth <= stack.last().unwrap().0 {
                stack.pop();
            }
            let new_node = Rc::new(RefCell::new(TreeNode::new(val)));
            let last_layer = stack.last().unwrap();
            if last_layer.1.borrow().left.is_none() {
                last_layer.1.borrow_mut().left = Some(new_node.clone());
            } else {
                last_layer.1.borrow_mut().right = Some(new_node.clone());
            }
            stack.push((depth, new_node));
        }

        let root = super_root.borrow_mut().left.clone();
        root
    }
}
// @lc code=end
#[test]
fn test_solution() {
    macro_rules! test {
        ($s:expr, $ans:tt) => {
            assert_eq!(
                Solution::recover_from_preorder($s.to_string()),
                btree!($ans)
            );
        };
    };
    test!("1-2--3--4-5--6--7", [1, 2, 5, 3, 4, 6, 7]);
    test!(
        "1-2--3---4-5--6---7",
        [1, 2, 5, 3, null, 6, null, 4, null, 7]
    );
    test!("1-401--349---90--88", [1, 401, null, 349, 88, 90]);
}
